# Chapter 3 Excercises
Logistic and Probit Regression Models

## Question 1
For the following 2x2 table, determine the odds and the probabilities of marijuana use among males and females. Then compute the odds ratio of marijuana use that compare males to females.

| Marijuana Use | Male | Female |
|:-------------:|:----:|:------:|
| Yes | 10 | 6 |
| No | 30 | 34 |


We know that the probability of a thing is: $$\text{probability}=\frac{\text{events}}{\text{number of outcomes}}$$

Therefore the probability of marijuana use among males is: $$p_{\text{males}}=\frac{10}{40}=0.25$$

and, the probability of marijuana use among females is: $$p_{\text{females}}=\frac{6}{40}=0.15$$

We compute the _odds_ using the general formula: $$\text{odds}=\frac{p}{1-p}$$

Therefore the odds of marijuana use among males is: $$\text{odds}_{\text{males}}=\frac{0.25}{1-0.25}=\frac{1}{3}$$

and, the odds of marijuana use among femailes is: $$\text{odds}_{\text{females}}=\frac{0.15}{1-0.15}=\frac{9}{50}$$

We compute the _odds ratio_ using the forumla: $$OR_{\text{a vs b}} = \frac{\text{odds}_{\text{a}}}{\text{odds}_{\text{b}}}$$

Therefore the odds ratio of males to females for marijuana use is: $$OR_{\text{males vs females}}=\frac{\frac{0.25}{1-0.25}}{\frac{0.15}{1-0.15}}=1.887$$

## Question 2
The data file _Religion_ contain a variable labeled _relschol_, which indicates whether or not the survey respondent attends a secondary school affiliated with a religious institution. The variable is coded as 0 = no and 1 = yes, conduct the following analysis:

### Part A
Compute the overall odds and probability of attending a religious school.

In [93]:
# For All Maths
import numpy as np
import pandas as pd
from sas7bdat import SAS7BDAT
import statsmodels.api as sm
import scipy
from patsy import dmatrices

In [94]:
# Load the data for this chapter
with SAS7BDAT('religion.sas7bdat') as f:
    df = f.to_data_frame()
df.head()

ID  SEX  AGE  EDUC  INCOME  RELSCHOL  MARRIED  ATTEND  AGESQ  RACE
0   2    1   30     6      11         0        1       6    900     1
1   3    1   32     6       6         0        1       5   1024     1
2   4    1   51     2      11         0        1       2   2601     1
3   5    1   18     2       3         0        0       6    324     1
4   6    1   37     5       6         0        1       6   1369     1

In [95]:
# Get our counts so we can calculate the probability of attendance
df['RELSCHOL'].value_counts()

0    546
1     80
dtype: int64

We know that the probability of a thing is: $$\text{probability}=\frac{\text{events}}{\text{number of outcomes}}$$

Therefore the probability of attending a religious school is $$p_{\text{relschool}}=\frac{80}{626}=0.1278$$

We compute the _odds_ using the general formula: $$\text{odds}=\frac{p}{1-p}$$

Therefore the odds of attending a religious school is $$\text{odds}_{\text{relschool}}=\frac{0.1278}{1-0.1278}=\frac{3}{20}$$

### Part B
Cross-tabulate _relschol_ with _race_ (coded 0 as non-write, 1 as white). What are the probabilities that non-white students and white students attend religious schools? What are the odds that white students and non-white students attend religious schools? What is the odds ratio that compares white and non-white students?

In [96]:
pd.crosstab(df['RELSCHOL'], df['RACE'])

RACE       0    1
RELSCHOL         
0         76  470
1         26   54

The probability of white students attending religious school is: $$p_{\text{white relschool}}=\frac{54}{524}=0.1031$$

The probability of non-white students attending religious school is: $$p_{\text{non-white relschool}}=\frac{26}{102}=0.2549$$

The odds of white students attending religious school is: $$\text{odds}_{\text{white relschool}}=\frac{0.1031}{1-0.1031}=\frac{11}{100}$$

The odds of non-white students attending religious school is: $$\text{odds}_{\text{non-white relschool}}=\frac{0.2549}{1-0.2549}=\frac{17}{50}$$

The odds ratio comparing white and non-white students is: $$OR_{\text{white vs non-white, relschool}}=\frac{\frac{0.1031}{1-0.1031}}{\frac{0.2549}{1-0.2549}}=0.34$$

## Queston 3
Estimate two logistic regression models that are designed to preict _relschol_. In the first model include only the variable _race_. In the second model, include _Race_, _attend_ (religious service attendence), and _income_ (family income), treating the latter two as continuous variables. Answer the following questions:

In [97]:
y, X = dmatrices('RELSCHOL ~ C(RACE)', data=df, return_type='dataframe')
model = sm.Logit(y, X)
results = model.fit()
print results.summary()

Optimization terminated successfully.
         Current function value: 0.370181
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:               RELSCHOL   No. Observations:                  626
Model:                          Logit   Df Residuals:                      624
Method:                           MLE   Df Model:                            1
Date:                Mon, 07 Sep 2015   Pseudo R-squ.:                 0.03138
Time:                        11:59:28   Log-Likelihood:                -231.73
converged:                       True   LL-Null:                       -239.24
                                        LLR p-value:                 0.0001066
                     coef    std err          z      P>|z|      [95.0% Conf. Int.]
----------------------------------------------------------------------------------
Intercept         -1.0726      0.227     -4.721      0.000        -1.518    -0.627
C(RACE)[T.1.0]   

In [98]:
print np.exp(results.params)

Intercept         0.342105
C(RACE)[T.1.0]    0.335843
dtype: float64


### Part A
Based on the first model, what is the odds ratio that compares white and non-white students? Compare this to the odds ratio computed in Excercise 2.B.

We know that we can take the exponent of the coefficient to acquire the odds ratio, in this case that is 0.3358. which is quite close to our calculated result in 2.B (0.336015546).

In [99]:
y, X = dmatrices('RELSCHOL ~ C(RACE)', data=df, return_type='dataframe')
model = sm.Logit(y, X)
results = model.fit()
print results.summary2()

Optimization terminated successfully.
         Current function value: 0.370181
         Iterations 6
                         Results: Logit
Model:              Logit            Pseudo R-squared: 0.031     
Dependent Variable: RELSCHOL         AIC:              467.4662  
Date:               2015-09-07 11:59 BIC:              476.3449  
No. Observations:   626              Log-Likelihood:   -231.73   
Df Model:           1                LL-Null:          -239.24   
Df Residuals:       624              LLR p-value:      0.00010659
Converged:          1.0000           Scale:            1.0000    
No. Iterations:     6.0000                                       
-----------------------------------------------------------------
                   Coef.  Std.Err.    z    P>|z|   [0.025  0.975]
-----------------------------------------------------------------
Intercept         -1.0726   0.2272 -4.7211 0.0000 -1.5179 -0.6273
C(RACE)[T.1.0]    -1.0911   0.2688 -4.0589 0.0000 -1.6180 -0.5642


In [100]:
y, X = dmatrices('RELSCHOL ~ C(RACE) + ATTEND + INCOME', data=df, return_type='dataframe')
model = sm.Logit(y, X)
results = model.fit()
print results.summary2()

Optimization terminated successfully.
         Current function value: 0.353214
         Iterations 7
                         Results: Logit
Model:              Logit            Pseudo R-squared: 0.080     
Dependent Variable: RELSCHOL         AIC:              424.7930  
Date:               2015-09-07 11:59 BIC:              442.3135  
No. Observations:   590              Log-Likelihood:   -208.40   
Df Model:           3                LL-Null:          -226.63   
Df Residuals:       586              LLR p-value:      5.9434e-08
Converged:          1.0000           Scale:            1.0000    
No. Iterations:     7.0000                                       
-----------------------------------------------------------------
                   Coef.  Std.Err.    z    P>|z|   [0.025  0.975]
-----------------------------------------------------------------
Intercept         -3.5831   0.7185 -4.9868 0.0000 -4.9914 -2.1749
C(RACE)[T.1.0]    -1.2893   0.2898 -4.4488 0.0000 -1.8573 -0.7213


### Part B
What are the AIC and BIC for the two models? Based on these measures of fit, which model do you prefer?

|   | AIC | BIC |
|---|:---:|:---:|
| $logit[p(Y=1)]=\alpha+\beta_1\text{RACE}$ | 467.4662  | 476.3449 |
| $logit[p(Y=1)]=\alpha+\beta_1\text{RACE}+\beta_2\text{ATTEND}+\beta_3\text{INCOME}$ | 424.7930 | 442.3135 |

The usual rule of thumb is to compare the AIC and BIC values and choose the model in which the values are lower. In this case the model performance on other parameters (Psuedo $R^2$) indicates that we get _some_ more explanation of variability by choosing the second model. However the question asks us to choose on AIC and BIC alone, therefor we would choose the second model.

### Part C
For tose who attend religious services five days per month (_attend_ = 5) and have a family income of $20,000-29,999$ (_income_ = 4), what are the predicted odds of attending a religious school for white and non-white students?

In [101]:
odds_white  = np.exp(-3.5831 + 1.0*-1.2893 + 5.0*0.3316 + 4.0*0.2007)
odds_non_white = np.exp(-3.5831 + 0.0*-1.2893 + 5.0*0.3316 + 4.0*0.2007)
print odds_white, odds_non_white

0.0896717050032 0.325530213444


The predicted odds of attending a religious school for those who are white, attend religious services 5 days per month, and have a family income of 20,000-29,999 is 0.08. The predicted odds of attending a religious school for those who are non-white, attend religious serivces 5 days per month, and have a family income of 20,000-29,999 is 0.32.

### Part D
What is the adjusted odds ratio for _race_? Interpret this odds ratio.

We will calculate the adjusted odds ratio for gender by plugging the mean values in and predicting the odds for white and non-white.

In [102]:
df.mean()

ID           503.487220
SEX            0.584665
AGE           47.077047
EDUC           3.323718
INCOME         5.215254
RELSCHOL       0.127796
MARRIED        0.635783
ATTEND         4.535144
AGESQ       2518.495987
RACE           0.837061
dtype: float64

In [103]:
m_odds_white  =np.exp(-3.5831 + 1.0*-1.2893 + 4.535144*0.3316 + 5.215254*0.2007)
m_odds_non_white = np.exp(-3.5831 + 0.0*-1.2893 + 4.535144*0.3316 + 5.215254*0.2007)
print m_odds_white / m_odds_non_white , np.exp(-1.2893)

0.275463540095 0.275463540095


We see that the odds ratio that we calculate from the mean values matches that of the exponiated coefficient that was calculated by our model. Thus for white attending religious school is about 1 in 4.

## Question 4
Re-estimate the two models outlined in Excercise 3, but use a probit model. Answer the following questions:

In [104]:
y, X = dmatrices('RELSCHOL ~ C(RACE)', data=df, return_type='dataframe')
model = sm.Probit(y, X)
results = model.fit()
print results.summary2()

Optimization terminated successfully.
         Current function value: 0.370181
         Iterations 6
                         Results: Probit
Model:              Probit           Pseudo R-squared: 0.031     
Dependent Variable: RELSCHOL         AIC:              467.4662  
Date:               2015-09-07 11:59 BIC:              476.3449  
No. Observations:   626              Log-Likelihood:   -231.73   
Df Model:           1                LL-Null:          -239.24   
Df Residuals:       624              LLR p-value:      0.00010659
Converged:          1.0000           Scale:            1.0000    
No. Iterations:     6.0000                                       
-----------------------------------------------------------------
                   Coef.  Std.Err.    z    P>|z|   [0.025  0.975]
-----------------------------------------------------------------
Intercept         -0.6591   0.1344 -4.9040 0.0000 -0.9226 -0.3957
C(RACE)[T.1.0]    -0.6052   0.1535 -3.9439 0.0001 -0.9060 -0.3044

### Part A
Based on the first model, what is the predicted probability that white and non-white students attend a religious school? Compare these results to those found in Exercise 2.B.

In [105]:
p_white = scipy.stats.norm.cdf(-0.6591-0.6051)
p_non_white = scipy.stats.norm.cdf(-0.6591)
print p_white, p_non_white

0.103079125424 0.25491577784


We see that the probabilities match what we calculated from the table in 2.B.

### Part B
What are the AIC and BIC for the two models? Compare these to the AIC and BIC computed in Exercise 3.B.


In [106]:
print results.summary2()

                         Results: Probit
Model:              Probit           Pseudo R-squared: 0.031     
Dependent Variable: RELSCHOL         AIC:              467.4662  
Date:               2015-09-07 11:59 BIC:              476.3449  
No. Observations:   626              Log-Likelihood:   -231.73   
Df Model:           1                LL-Null:          -239.24   
Df Residuals:       624              LLR p-value:      0.00010659
Converged:          1.0000           Scale:            1.0000    
No. Iterations:     6.0000                                       
-----------------------------------------------------------------
                   Coef.  Std.Err.    z    P>|z|   [0.025  0.975]
-----------------------------------------------------------------
Intercept         -0.6591   0.1344 -4.9040 0.0000 -0.9226 -0.3957
C(RACE)[T.1.0]    -0.6052   0.1535 -3.9439 0.0001 -0.9060 -0.3044



In [107]:
y, X = dmatrices('RELSCHOL ~ C(RACE) + ATTEND + INCOME', data=df, return_type='dataframe')
model = sm.Probit(y, X)
results = model.fit()
print results.summary2()

Optimization terminated successfully.
         Current function value: 0.351750
         Iterations 6
                         Results: Probit
Model:              Probit           Pseudo R-squared: 0.084     
Dependent Variable: RELSCHOL         AIC:              423.0652  
Date:               2015-09-07 11:59 BIC:              440.5857  
No. Observations:   590              Log-Likelihood:   -207.53   
Df Model:           3                LL-Null:          -226.63   
Df Residuals:       586              LLR p-value:      2.5609e-08
Converged:          1.0000           Scale:            1.0000    
No. Iterations:     6.0000                                       
-----------------------------------------------------------------
                   Coef.  Std.Err.    z    P>|z|   [0.025  0.975]
-----------------------------------------------------------------
Intercept         -2.0718   0.3849 -5.3824 0.0000 -2.8263 -1.3174
C(RACE)[T.1.0]    -0.7261   0.1631 -4.4527 0.0000 -1.0457 -0.4065

|   | AIC | BIC |
|---|:---:|:---:|
| $logit[p(Y=1)]=\alpha+\beta_1\text{RACE}$ | 467.4662  | 476.3449 |
| $logit[p(Y=1)]=\alpha+\beta_1\text{RACE}+\beta_2\text{ATTEND}+\beta_3\text{INCOME}$ | 424.7930 | 442.3135 |
| $probit[p(Y=1)]=\alpha+\beta_1\text{RACE}$ | 467.4662 | 476.3449 |
| $probit[p(Y=1)]=\alpha+\beta_1\text{RACE}+\beta_2\text{ATTEND}+\beta_3\text{INCOME}$ | 423.0652 | 440.5857 |

We see that the AIC and BIC between Logit and Probit models is very similar. We would opt to choose the second model once more due to its comparatively lower AIC and BIC values.

### Part C
For tose who attend religious services five days per month (_attend_ = 5) and have a family income of $20,000-29,999$ (_income_ = 4), what are the predicted probabilities of attending a religious school for white and non-white students?

In [108]:
p_white  = scipy.stats.norm.cdf(-2.0718 + 1.0*-0.7261 + 5.0*0.1858 + 4.0*0.1156)
p_non_white = scipy.stats.norm.cdf(-2.0718 + 0.0*-0.7261 + 5.0*0.1858 + 4.0*0.1156)
print odds_white, odds_non_white

0.0896717050032 0.325530213444


### Part D
Compute the discrete change in probability under the following scenario: A non-white student whose _attend_ value equals 4 with a shift in family income (_income_) rom a value of 4($20,000-29,999$) to a value of 10 ($80,000-99,999$)

For continuous variables the discrete change is harder to compute, but the interpretation is similar. The following forula is used for computing the discrete changes: $$\frac{\Delta P(y=1|\bar{x})}{\Delta x_i}=P(y=1|\mathbf{\bar{x}},\bar{x_k}+1)-P(y=1|\mathbf{\bar{x}},\bar{x_k})$$

In [109]:
p_4 = scipy.stats.norm.cdf(-2.0718 + 0.0*-0.7261 + 5.0*0.1858 + 4.0*0.1156)
p_10 = scipy.stats.norm.cdf(-2.0718 + 0.0*-0.7261 + 5.0*0.1858 + 10.0*0.1156)
p_4 - p_10

-0.25714027466459394

## Question 5
In plain English, what do you conclude about the relationship between a student's race/ehnicity, religious service attendance, family income, and attending a religious school?

Each have an impact on the probability that a student will attend a religious school. Each, from the data, can be teased out so that we can have some indication of it's descrete effect. 